docs/specs/kemsphinx.rst

```diff
@@ -30,7 +32,69 @@ has two functions:
 Therefore we must embed these KEM ciphertexts in the KEMSphinx header,
 one KEM ciphertext per mix hop.
 
-2. KEMSphinx Header Design
+2. Post Quantum Hybrid KEM
+==========================
+
+Special care must be taken in order correctly compose a hybrid post
+quantum KEM that is IND-CCA2 robust.
+
+The hybrid post quantum KEMs found in Cloudflare's circl library are
+suitable to be used with Noise or TLS but not with KEM Sphinx because
+they are not IND-CCA2 robust. Noise and TLS achieve IND-CCA2 security
+by mixing in the public keys and ciphertexts into the hash object and
+therefore do not require an IND-CCA2 KEM.
+
+Firstly, our post quantum KEM is IND-CCA2 however we must specifically take care
+to make our NIKE to KEM adapter have semantic security. Secondly, we must make a
+security preserving KEM combiner.
+
+2.1 NIKE to KEM adapter
+-----------------------
+
+We easily achieve our IND-CCA2 security by means of hashing
+together the DH shared secret along with both of the public keys:
+
+```
+func Encap(their_pubkey publickey) ([]byte, []byte) {
+        my_privkey, my_pubKey = GEN_KEYPAIR(RNG)
+        ss = DH(my_privkey, their_pubkey)
+        ss2 = H(ss || my_pubkey || their_pubkey)
+	return my_pubkey, ss2
+}
+
+func Decap(my_privkey, ciphertext) []byte {
+        s = DH(my_privkey, ciphertext)
+	shared_key = H(ss || my_pubkey || ciphertext)
+	return shared_eky
+}
+```
+
+
+2.2 KEM Combiner
+----------------
+
+The KEM Combiners paper [KEMCOMB]_ makes the observation that if a KEM combiner
+is not security preserving then the resulting hybrid KEM will not have
+IND-CCA2 security if one of the composing KEMs does not have IND-CCA2
+security. Likewise the paper points out that when using a security
+preserving KEM combiner, if only one of the composing KEMs has
+IND-CCA2 security then the resulting hybrid KEM will have IND-CCA2
+security.
+
+Our KEM combiner uses the split PRF design from the paper when combining
+two KEM shared secrets together we use a hash function to also mix
+in the values of both KEM ciphertexts:
+
+```
+// H is a hash function
+
+func splitPRF(ss1, ss2, cct1, cct2 []byte) []byte {
+        return H(ss1 || ss2 || cct1 || cct2)
+}
+```
```

Comment:
This seems good, but `cct1` and `cct2` should be defined explicitly in `Encap`/`Decap`: it should return a `cct` value, the concatenation of the public keys in a defined order; that way the combiner doesn't have to violate the layers.

Reply:
For this to be a PRF I believe it would need to take an additional parameter to parameterize the session key over?
Like we'd want to use it something like
 `recv_key, send_key := splitPRF(blabla, "server"), splitPRF(blabla, "client")`

Reply:
no... this is perfectly correct as is. let's talk elsewhere or have a phone call or something.